File: app/src/main/java/com/jim/dagger2java/MainActivity.java
package com.jim.dagger2java;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import com.jim.dagger2java.car.Car;
import com.jim.dagger2java.dagger.CarComponent;
import com.jim.dagger2java.dagger.DaggerCarComponent;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity {

    @Inject
    Car car;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        CarComponent component = DaggerCarComponent.builder()
                .horsePower(100)
                .engineCapacity(1400)
                .build();
        component.inject(this);
        car.drive();
    }
}

File: Retrofit-With-Hilt/src/main/java/com/app/hiltretrofit/main/view/MainActivity.kt
package com.app.hiltretrofit.main.view

import android.os.Bundle
import android.widget.LinearLayout
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.app.hiltretrofit.databinding.ActivityMainBinding
import com.app.hiltretrofit.main.view.adapter.CountryAdapter
import com.app.hiltretrofit.main.viewmodel.MainViewModel
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private val mainViewModel: MainViewModel by viewModels()

    //inject the country Adaptor
    @Inject
    lateinit var countryAdapter: CountryAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setUpUI()
        setUpObservers()
    }

    private fun setUpUI() {
        binding.countryRecyclerview.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            setHasFixedSize(true)
            adapter = countryAdapter
            addItemDecoration(DividerItemDecoration(this@MainActivity, LinearLayout.VERTICAL))
        }
    }

    private fun setUpObservers() {
        mainViewModel.getCountry().observe(this, { countryList ->
            countryList?.let {
                binding.countryRecyclerview.apply {
                    with(countryAdapter) {
                        countries = it
                        notifyDataSetChanged()
                    }
                }
            }
        })
    }
}
